src/libraries/System.Private.CoreLib/src/System/Buffer.cs

```diff
@@ -350,11 +350,33 @@ internal static unsafe void Memmove<T>(ref T destination, ref T source, nuint el
 #pragma warning disable 8500 // sizeof of managed types
             if (!RuntimeHelpers.IsReferenceOrContainsReferences<T>())
             {
+                ref byte dst = ref Unsafe.As<T, byte>(ref destination);
+                ref byte src = ref Unsafe.As<T, byte>(ref source);
+                nuint byteLen = elementCount * (nuint)sizeof(T);
+
+#if TARGET_64BIT
+                // If pointers don't overlap and length is small enough -
+                // we can unroll it with Unsafe.CopyBlockUnaligned.
+                // 128 is based on JIT's internal getUnrollThreshold(memcpy) implementation, keep in sync
+                const nuint unrollThreshold = 128;
+                if (RuntimeHelpers.IsKnownConstant(byteLen) && byteLen <= unrollThreshold)
+                {
+                    // This check can be made faster with abs()
+                    if (((nuint)(nint)Unsafe.ByteOffset(ref src, ref dst) < byteLen) ||
```

Comment:
Copied from https://github.com/dotnet/runtime/blob/4a1dd91fe88e59338072056afdbf3f7a6b426eb6/src/libraries/System.Private.CoreLib/src/System/Buffer.cs#L130-L133

Reply:
I think it would be better to expand the whole thing as intrinsic in the JIT instead of trying to handle it partially here.

The JIT should be able to prove that the buffers are non-overlapping in many cases. For example when you are copying from a string or a constant RVA block, it is ok for the JIT to assume that the buffers are non-overlapping. Or when you are copying from a stack to GC heap, or vice versa, it is safe to assume that the buffers are non-overlapping.

When the buffers are potentially overlapping, you can still do inline expansion (with smaller threshold) by first reading the whole thing into registers and then writing the whole thing out.

Eventually, once the arm memory copy instruction become available, we will want to handle memmove as intrinsic and always expand it to the memory copy instructions, if no other optimizations kick in.

Reply:
FYI: LLVM unrolls even for potentially overlapping memory by reading it first and storing into regs.

Reply:
The issue here would be in figuring out how many registers can be used for this.